BlueprintUI/Sources/Internal/ElementIdentifier.swift

```diff
@@ -1,43 +1,114 @@
 
-    /// The element represented by this component was not assigned a specific key, so its index in the parent's array
-    /// of children is used instead.
-    case index(Int)
+ A fully constructed `ElementIdentifier` would look like this:
+ 
+ ```
+ ElementIdentifier(elementType: MyElement.self, key: nil, count: 2)
+ ```
+ 
+ Which means that this identifier represents an element of type `MyElement`, which is
+ the second `MyElement` element in the hierarchy.
+ 
+ Elements are identified and counted by their `elementType` and `key` to further assist in diffing during update operations.
+ 
+ For example, if we started with:
+ ```
+ ElementA -> ElementIdentifier(ElementA.self, [optional key], 1)
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 1)
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 2)
+ ElementC -> ElementIdentifier(ElementC.self, [optional key], 1)
+ ```
+ 
+ And then removed A and the first B:
+ 
+ ```
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 1)
+ ElementC -> ElementIdentifier(ElementC.self, [optional key], 1)
+ ```
+ 
+ You will note that the identifiers remain stable, which ultimately ensures that views are reused.
+ */
+struct ElementIdentifier: Hashable, CustomDebugStringConvertible {
+    
+    let elementType : ObjectIdentifier
+    let key : AnyHashable?
 
-    internal init(key: String?, index: Int) {
-        if let key = key {
-            self = .key(key)
+    let count : Int
+
+    init(elementType : Element.Type, key : AnyHashable?, count : Int)
+    {
+        self.elementType = ObjectIdentifier(elementType)
+        self.key = key
+        
+        self.count = count
+    }
+    
+    var debugDescription: String {
+        if let key = self.key {
+            return "\(self.elementType).\(String(describing: key)).\(self.count)"
         } else {
-            self = .index(index)
+            return "\(self.elementType).\(self.count)"
         }
     }
-
-    /// Returns the reuse identifier of this component (if provided).
-    var key: String? {
-        switch self {
-        case .key(let key):
-            return key
-        default:
-            return nil
+    
+    /**
+     Internal type used to create `ElementIdentifier` instances during view hierarchy updates.
+     */
+    struct Factory
+    {
+        init(reserveCapacity capacity: Int? = nil)
+        {
+            self.countsByKey = [:]
+            
+            if let capacity = capacity {
+                self.countsByKey.reserveCapacity(capacity)
+            }
         }
-    }
-
-}
-
-extension ElementIdentifier: CustomStringConvertible {
-
-    var description: String {
-        switch self {
-        case let .key(string):
-            return string
-        case let .index(index):
-            return "(\(index))"
+                
+        mutating func nextIdentifier(for type : Element.Type, key : AnyHashable?) -> ElementIdentifier
```

Comment:
This type has a lot of stylistic differences from the rest of the project, i.e., spaces before colons and the return before brackets. Should we set up Swiftlint or something for Blueprint?